ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -139,3 +152,81 @@ muxChannel pmss mid md w cnt =
+            Ord ptcl, Enum ptcl, Bounded ptcl, HasCallStack)
+        => [(Version, MiniProtocolDescriptions ptcl m)]
+        -> MuxBearer ptcl m
+        -> m (Version, MiniProtocolDescriptions ptcl m)
+muxClient verMpds bearer = do
+    let versions = map fst verMpds
+        msg = MsgInitReq $ M.fromList $ map (\v -> (versionToVersionNumber v, v)) versions
+        blob = toLazyByteString $ encodeCtrlMsg msg
+        sdu = MuxSDU (RemoteClockModel 0) Muxcontrol ModeInitiator (fromIntegral $ BL.length blob) blob
+
+    void $ write bearer sdu
+    (rsp, _) <- Ouroboros.Network.Mux.Types.read bearer
+    if msId rsp /= Muxcontrol || msMode rsp /= ModeResponder
+       then throwM $ MuxError MuxUnknownMiniProtocol "invalid muxInit rsp id or mode" callStack
+       else do
+           let rspMsg_e = CBOR.deserialiseFromBytes decodeCtrlMsg (msBlob rsp)
+           case rspMsg_e of
+                Left e -> throwM e
+                Right (_, MsgInitRsp version) ->
+                    -- verify that rsp version matches one of our proposals
+                    case lookup version verMpds of
+                         Nothing   -> throwM $ MuxError MuxControlProtocolError
+                                               "muxInit invalid version selected" callStack
+                         Just mpds -> return (version, mpds)
+                Right (_, MsgInitReq _) -> throwM $ MuxError MuxControlProtocolError
+                                                    "muxInit response as request" callStack
+                Right (_, MsgInitFail e) -> throwM $ MuxError MuxControlNoMatchingVersion e callStack
+
+-- | Wait for the connected peer to initiate version negotiation.
+muxServer :: (MonadAsync m, MonadFork m, MonadSay m, MonadSTM m, MonadThrow m,
+            Ord ptcl, Enum ptcl, Bounded ptcl, HasCallStack)
+        => [(Version, MiniProtocolDescriptions ptcl m)]
+        -> MuxBearer ptcl m
+        -> m (Version, MiniProtocolDescriptions ptcl m)
+muxServer verMpds bearer = do
+    let localVersions = map fst verMpds
+    (req, _) <- Ouroboros.Network.Mux.Types.read bearer
+    if msId req /= Muxcontrol || msMode req /= ModeInitiator
+       then throwM $ MuxError MuxUnknownMiniProtocol "invalid muxInit req id or mode" callStack
+       else do
+           let reqMsg_e = CBOR.deserialiseFromBytes decodeCtrlMsg (msBlob req)
+           case reqMsg_e of
+                Left e -> throwM e
+                Right (_, MsgInitReq remoteVersionsM) -> do
+                    let remoteVersions = M.elems remoteVersionsM
+                        matchingVersions = localVersions `intersect` remoteVersions
+                    if null matchingVersions
+                       then do
+                           sndSdu $ MsgInitFail $ "No matching version, try " ++
+                                show (last localVersions)
+
+                           throwM $ MuxError MuxControlNoMatchingVersion "no matching version" callStack
```

Comment:
Could you document that `muxStart` can raise this error. It would be good to list all the errors `muxStart` can throw.  In the future this one can be handled by the caller: we know that this peer is running incompatible versions and we should not waste the time to contact it, at least for some time. Later we might want to know if there is a more recent version of the protocol, that we should use (`Ord` instance on `Version` might be useful).

Reply:
This is the passive part, they are contacting us not the other way around.